lib/shrine.rb

```diff
@@ -799,9 +799,14 @@ def download(*args)
           if storage.respond_to?(:download)
             storage.download(id, *args)
           else
-            tempfile = Tempfile.new(["shrine", ".#{extension}"], binmode: true)
-            open(*args) { |io| IO.copy_stream(io, tempfile.path) }
-            tempfile.tap(&:open)
+            begin
+              tempfile = Tempfile.new(["shrine", ".#{extension}"], binmode: true)
+              open(*args) { |io| IO.copy_stream(io, tempfile.path) }
+              tempfile.tap(&:open)
+            rescue => err
+              tempfile.close!
```

Comment:
If an exception occurs in `Tempfile.new` (e.g. there in case there are too many open file descriptors, so Ruby cannot create more), then the `tempfile` variable will be nil. We should handle that by closing the `tempfile` only if it was actually created:

```rb
tempfile.close! if tempfile
```

We should also test this scenario.

Reply:
Good to know about the too many open files limitation. Agree.